pypeit/coadd3d.py

```diff
@@ -0,0 +1,1853 @@
+            else:
+                msgs.warn(f'Processed alignment frame not recorded or not found!')
+                msgs.info("Using slit edges for astrometric transform")
+        else:
+            msgs.info("Using slit edges for astrometric transform")
+        # If nothing better was provided, use the slit edges
+        if alignments is None:
+            left, right, _ = slits.select_edges(initial=True, flexure=spat_flexure)
+            locations = [0.0, 1.0]
+            traces = np.append(left[:, None, :], right[:, None, :], axis=1)
+        else:
+            locations = self.par['calibrations']['alignment']['locations']
+            traces = alignments.traces
+        msgs.info("Generating alignment splines")
+        return alignframe.AlignmentSplines(traces, locations, spec2DObj.tilts)
+
+    def set_voxel_sampling(self):
+        """
+        This function checks if the spatial and spectral scales of all frames are consistent.
+        If the user has not specified either the spatial or spectral scales, they will be set here.
+        """
+        # Make sure all frames have consistent pixel scales
+        ratio = (self._spatscale[:, 0] - self._spatscale[0, 0]) / self._spatscale[0, 0]
+        if np.any(np.abs(ratio) > 1E-4):
+            msgs.warn("The pixel scales of all input frames are not the same!")
+            spatstr = ", ".join(["{0:.6f}".format(ss) for ss in self._spatscale[:,0]*3600.0])
+            msgs.info("Pixel scales of all input frames:" + msgs.newline() + spatstr + "arcseconds")
+        # Make sure all frames have consistent slicer scales
+        ratio = (self._spatscale[:, 1] - self._spatscale[0, 1]) / self._spatscale[0, 1]
+        if np.any(np.abs(ratio) > 1E-4):
+            msgs.warn("The slicer scales of all input frames are not the same!")
+            spatstr = ", ".join(["{0:.6f}".format(ss) for ss in self._spatscale[:,1]*3600.0])
+            msgs.info("Slicer scales of all input frames:" + msgs.newline() + spatstr + "arcseconds")
+        # Make sure all frames have consistent wavelength sampling
+        ratio = (self._specscale - self._specscale[0]) / self._specscale[0]
+        if np.any(np.abs(ratio) > 1E-2):
+            msgs.warn("The wavelength samplings of the input frames are not the same!")
+            specstr = ", ".join(["{0:.6f}".format(ss) for ss in self._specscale])
+            msgs.info("Wavelength samplings of all input frames:" + msgs.newline() + specstr)
+
+        # If the user has not specified the spatial scale, then set it appropriately now to the largest spatial scale
+        if self._dspat is None:
+            self._dspat = np.max(self._spatscale)
+            msgs.info("Adopting a square pixel spatial scale of {0:f} arcsec".format(3600.0 * self._dspat))
+        # If the user has not specified the spectral sampling, then set it now to the largest value
+        if self._dwv is None:
+            self._dwv = np.max(self._specscale)
```

Comment:
Can you return these values instead of assigning to self. For a small number of return values that makes the control flow more transparent.